### DISCUSSION TITLE
Middleware doesn't support basePath option

### DISCUSSION THREAD

**Discussion by lavcraft** [nextauthjs/next-auth#5502] [Ideas]
### Environment

next-auth: 4.12.2
next: 12.3.1

### Reproduction URL

https://github.com/lavcraft/next-auth-basepath-bug

### Describe the issue

This bug trigger when you use basePath option in next.js and configure it via next-auth SessionProvider (basePath property)

After call [handleMiddleware function](https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/next/middleware.ts#L149) in next-auth package, return value with target callbackUrl doesn't polite with basePath

In this line:
```typescript
signInUrl.searchParams.append("callbackUrl", `${pathname}${search}`)
```
I this, this will be fixed with simple hotfix

```typescript
signInUrl.searchParams.append("callbackUrl", `${req.nextUrl.basePath ?? ''}${pathname}${search}`)
```

### How to reproduce

1. Clone repo https://github.com/lavcraft/next-auth-basepath-bug
2. run `npm ci`
3. run `npm run dev`
4. open http://localhost:3001/custom/me
5. see invalid callbackUrl in query param (without /custom basePath)

### Expected behavior

Stored callbackUrl include basePath for handle it correctly after authentication process or in other cases

PS: May be default value for pages: { .. } in next-auth middleware withAuth args also support basePath?
```typescript
import { withAuth } from "next-auth/middleware"

const basePath = process.env.NEXT_PATH_PREFIX ?? '';
// More on how NextAuth.js middleware works: https://next-auth.js.org/configuration/nextjs#middleware
export default withAuth({
  callbacks: {
    ....
  },
  pages: {
    signIn: `${basePath}/api/auth/signin` //<-- i mean this, avoid basePath for defaultValue
  },
})

export const config = { matcher: ["/admin", "/me"] }
```

**Comment by ThangHuuVu** [comment]
There is an ongoing PR to fix this 👀 https://github.com/nextauthjs/next-auth/pull/5109 It has a different approach than yours. I suggest we continue in #5109 and work together to find the best solution there! 🙌 Let's keep this issue open in the meantime.

**Comment by lavcraft** [comment]
@ThangHuuVu thanks. I have looked in details your PR. I this you follow very similar approach. Only one major difference - 

1. i'm using basePath field from next.js object, 
2. You trying construct basePath by you own util function

~~And both implementation has a problem with URL with templated args like /order/**[orderId]**/page i think :) But i advice to solve this in next issues~~ - sorry, check my solution, its work correctly in this case :)
